drivers/rtc/rtc_rv3028.c

```diff
@@ -0,0 +1,1101 @@
+{
+	struct rv3028_data *data = dev->data;
+	uint8_t date[7];
+	int err;
+
+	if (timeptr == NULL ||
+	    (timeptr->tm_year < RV3028_YEAR_OFFSET) ||
+	    (timeptr->tm_year > (RV3028_YEAR_OFFSET + 99))) {
+		LOG_ERR("Invalid time");
+		return -EINVAL;
+	}
+
+	k_mutex_lock(&data->lock, K_FOREVER);
+
+	LOG_DBG("Set time: year = %d, mon = %d, mday = %d, wday = %d, hour = %d, "
+	        "min = %d, sec = %d",
+	        timeptr->tm_year, timeptr->tm_mon, timeptr->tm_mday, timeptr->tm_wday,
+	        timeptr->tm_hour, timeptr->tm_min, timeptr->tm_sec);
+
+	date[0] = bin2bcd(timeptr->tm_sec) & RV3028_SECONDS_MASK;
+	date[1] = bin2bcd(timeptr->tm_min) & RV3028_MINUTES_MASK;
+	date[2] = bin2bcd(timeptr->tm_hour) & RV3028_HOURS_24H_MASK;
+	date[3] = bin2bcd(timeptr->tm_wday) & RV3028_WEEKDAY_MASK;
+	date[4] = bin2bcd(timeptr->tm_mday) & RV3028_DATE_MASK;
+	date[5] = bin2bcd(timeptr->tm_mon + RV3028_MONTH_OFFSET) & RV3028_MONTH_MASK;
+	date[6] = bin2bcd(timeptr->tm_year - RV3028_YEAR_OFFSET) & RV3028_YEAR_MASK;
+
+	err = rv3028_write_regs(dev, RV3028_REG_SECONDS, &date, sizeof(date));
+
+	k_mutex_unlock(&data->lock);
+
+	return err;
+}
+
+static int rv3028_get_time(const struct device *dev, struct rtc_time *timeptr)
+{
+	uint8_t date[7];
+	int err;
+
+	if (timeptr == NULL) {
+		return -EINVAL;
+	}
+
+	err = rv3028_read_regs(dev, RV3028_REG_SECONDS, date, sizeof(date));
+	if (err) {
+		return err;
+	}
+
+	memset(timeptr, 0U, sizeof(*timeptr));
+	timeptr->tm_sec = bcd2bin(date[0] & RV3028_SECONDS_MASK);
+	timeptr->tm_min = bcd2bin(date[1] & RV3028_MINUTES_MASK);
+	timeptr->tm_hour = bcd2bin(date[2] & RV3028_HOURS_24H_MASK);
+	timeptr->tm_wday = bcd2bin(date[3] & RV3028_WEEKDAY_MASK);
+	timeptr->tm_mday = bcd2bin(date[4] & RV3028_DATE_MASK);
+	timeptr->tm_mon = bcd2bin(date[5] & RV3028_MONTH_MASK) - RV3028_MONTH_OFFSET;
+	timeptr->tm_year = bcd2bin(date[6] & RV3028_YEAR_MASK) + RV3028_YEAR_OFFSET;
+	timeptr->tm_yday = -1;
+	timeptr->tm_isdst = -1;
+
+	LOG_DBG("Get time: year = %d, mon = %d, mday = %d, wday = %d, hour = %d, "
+	        "min = %d, sec = %d",
+	        timeptr->tm_year, timeptr->tm_mon, timeptr->tm_mday, timeptr->tm_wday,
+	        timeptr->tm_hour, timeptr->tm_min, timeptr->tm_sec);
+
+	return 0;
+}
+
+#ifdef CONFIG_RTC_ALARM
+
+static int rv3028_alarm_get_supported_fields(const struct device *dev, uint16_t id, uint16_t *mask)
+{
+	ARG_UNUSED(dev);
+
+	if (id != 0U) {
+		LOG_ERR("Invalid alarm ID %d", id);
+		return -EINVAL;
+	}
+
+	*mask = RV3028_RTC_ALARM_TIME_MASK;
+
+	return 0;
+}
+
+static int rv3028_alarm_set_time(const struct device *dev, uint16_t id, uint16_t mask,
+                                 const struct rtc_time *timeptr)
+{
+	uint8_t regs[3];
+
+	if (id != 0U) {
+		LOG_ERR("Invalid alarm ID %d", id);
+		return -EINVAL;
+	}
+
+	if ((mask & ~(RV3028_RTC_ALARM_TIME_MASK)) != 0U) {
+		LOG_ERR("Unsupported alarm field mask 0x%04x", mask);
+		return -EINVAL;
+	}
+
```

Comment:
Please validate the timeptr using https://github.com/zephyrproject-rtos/zephyr/blob/3570408db556fbb53cd775ae3d2e44553e28cb31/drivers/rtc/rtc_utils.h#L26 here as well

Reply:
I've updated the time validation